### DISCUSSION TITLE
test_fund_me.py returns AttributeError: 'NoneType' object has no attribute 'getEntranceFee'

### DISCUSSION THREAD

**Discussion by klovertexarkana** [smartcontractkit/full-blockchain-solidity-course-py#1750] [Q&A] [answered]
I keep getting the above error. Prior to that testing returns:

The active network is development
Deploying Mocks...
Mocks deployed!
Contract deployed to 0x602C71e4DAC47a042Ee7f46E0aee17F94A3bA0B6

So it appear that my deploy script is working fine and I haven't had any issues deploying to development, ganache-local, or Goerli.

Here is test_fund_me.py:

```Python
from scripts.helpful_scripts import get_account
from scripts.deploy import deploy_fund_me


def test_can_fund_and_withdraw():
    account = get_account()
    fund_me = deploy_fund_me()
    entrance_fee = fund_me.getEntranceFee()

    tx = fund_me.fund({'from': account, 'value': entrance_fee})
    tx.wait(1)
    assert fund_me.addressToAmountFunded(account.address) == entrance_fee

    tx2 = fund_me.withdraw({'from': account})
    tx2.wait(1)
    assert fund_me.addressToAmountFunded(account.address) == 0

Here is my deploy.py:

from brownie import network, FundMe, MockV3Aggregator, config
from scripts.helpful_scripts import get_account, deploy_mocks, LOCAL_BLOCKCHAIN_ENVIRONMENTS


def deploy_fund_me():
    account = get_account()
    # need to pass appropriate price feed address to the .deploy() method
    if network.show_active() not in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        price_feed_address = config['networks'][network.show_active()]['eth_usd_price_feed']
    else:
        deploy_mocks()
        price_feed_address = MockV3Aggregator[-1].address


    fund_me = FundMe.deploy(price_feed_address, {'from': account},
                            publish_source=config['networks'][network.show_active()].get('verify'))
    print(f'Contract deployed to {fund_me.address}')


def main():
    deploy_fund_me()
```

**Comment by cromewar** [comment, marked as answer]
Hello @klovertexarkana this might happen to several reasons.
1. you should wait one block confirmation before calling the `getEntranceFee` to ensure the contract was deployed.
2. Check you contract, does the function is called exactly `getEntranceFee`? may be there is a typo causing the problem.
3. You might have a bad contract cache, so delete the build folder on your project and try again.

I hope any of this info might help.

**Reply by klovertexarkana** [reply]
Due to the fact that this is returned:  "AttributeError: 'NoneType' object has no attribute 'getEntranceFee'", but I get the print out that my contract was deployed successfully, I can't help but think that the contract gets deployed correctly and that for some reason brownie's test command is having a problem connecting to it. Hence, 'NoneType' object. Brownie is not able to utilize the getEntranceFee method because it has not contract to find the getEntranceFee method.


**1. you should wait one block confirmation before calling the getEntranceFee to ensure the contract was deployed.**

I tried the below to see if it would work and it did not:
```Python

from scripts.helpful_scripts import get_account
from scripts.deploy import deploy_fund_me


def test_can_fund_and_withdraw():
    account = get_account()
    fund_me = deploy_fund_me()
    fund_me.wait(1)
    entrance_fee = fund_me.getEntranceFee()

    tx = fund_me.fund({'from': account, 'value': entrance_fee})
    tx.wait(1)
    assert fund_me.addressToAmountFunded(account.address) == entrance_fee

    tx2 = fund_me.withdraw({'from': account})
    tx2.wait(1)
    assert fund_me.addressToAmountFunded(account.address) == 0

```
**2. Check you contract, does the function is called exactly getEntranceFee? may be there is a typo causing the problem.**

Here is my getEntranceFee() from FundMe.sol:

    function getEntranceFee() public view returns (uint256) {
        // minimumUSD
        uint256 minimumUSD = 50 * 10**18;
        uint256 price = getPrice();
        uint256 precision = 1 * 10**18;
        return (minimumUSD * precision) / price;

**3. You might have a bad contract cache, so delete the build folder on your project and try again.**

I deleted my build folder, re-compiled and deployed to 'ganache-local' and Goerlie and there was not problem at all. Testing still produced the same error output.

**Reply by cromewar** [reply]
Ok I found your problem on the deploy script `deploy_fund_me` you must return `fund_me` at the end, otherwise when you go `fund_me = deploy_fund_me()` it will be empty. Check it out the code repo: https://github.com/PatrickAlphaC/brownie_fund_me/blob/main/scripts/deploy.py

**Reply by klovertexarkana** [reply]
Wow, well that certainly makes sense! And yes it now works fine. Thank you!

**Reply by cromewar** [reply]
I'm glad is solved now.
